curp/src/server/cmd_board.rs

```diff
@@ -157,18 +169,18 @@ impl<C: Command> CommandBoard<C> {
     pub(super) async fn wait_for_er_asr(
         cb: &CmdBoardRef<C>,
         id: &ProposeId,
-    ) -> (Result<C::ER, C::Error>, Option<Result<C::ASR, C::Error>>) {
-        loop {
-            {
-                let cb_r = cb.read();
-                match (cb_r.er_buffer.get(id), cb_r.asr_buffer.get(id)) {
-                    (Some(er), None) if er.is_err() => return (er.clone(), None),
-                    (Some(er), Some(asr)) => return (er.clone(), Some(asr.clone())),
-                    _ => {}
-                }
-            }
-            let listener = cb.write().asr_listener(id);
-            listener.await;
+    ) -> Option<(Result<C::ER, C::Error>, Option<Result<C::ASR, C::Error>>)> {
+        match cb.map_read(|cb_r| cb_r.get_er_asr(id)) {
+            (Some(er), None) if er.is_err() => return Some((er, None)),
+            (Some(er), Some(asr)) => return Some((er, Some(asr))),
+            _ => {}
+        }
+        let listener = cb.write().asr_listener(id);
+        listener.await;
+        match cb.map_read(|cb_r| cb_r.get_er_asr(id)) {
+            (Some(er), None) if er.is_err() => Some((er, None)),
+            (Some(er), Some(asr)) => Some((er, Some(asr))),
+            _ => None,
         }
```

Comment:
Could we let `get_er_asr` return `Option<(Result<C::ER, C::Error>, Option<Result<C::ASR, C::Error>>)` directly?